### PR TITLE
all: fix all undefined doxygen groups

### LINKED ISSUE #8972: [Feature Request] doc: create top-level module checker
#### Description
It appears to be every week or so that someone breaks the [top-level module view](http://doc.riot-os.org/modules.html) of our nice rendered doc (see #8973 for the most current example). Main culprit most of the time is that they set `@ingroup` to a group that does not exist.

How about we provide a checker script that checks `modules.html` directly after the doc was build by the CI, if the list of modules is in line with a given list provided with that checker? If the author of a PR really intends to add a new top-level module (which should rarely happen anyway) they can just append there new module to that list.

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR fixes #8972 by adding a doxygen groups checker script. This script is based on the [gist](https://gist.github.com/kaspar030/78b6a26e710a03ae1d28cae5c38f3cbe) provided by @kaspar030 and which is working very well.
I slightly adapted it to not take into account the vendor header files and to make it return the full list of non matching groups.

In the meantime, I fixed all issues raised by the script. But I think I'll split this PR in several parts to simplify the reviews:
1. provide the check script alone, e.g not integrated in the CI
2. provide a PR per top folder (boards, cpu, etc)
3. enable the script in the CI

### Issues/PRs references

fixes #8972 ~and based on #9266~

### CODE REVIEW

**Comment by miri64** [comment]
Can we divide this up into two PRs please? One introducing the script (but not wired hot yet, just silently warning in the output) and one fixing the warnings?

Also, since this is doc, why isn't this included into `doccheck`?

**Comment by aabadie** [comment]
> Can we divide this up into two PRs please?

Sure, that's what I wrote in the initial comment of this PR.

> since this is doc, why isn't this included into doccheck?

I thought it would be simpler to have it separate first. But I'm fine with merging the two once all grouping issues are fixed.

File: doc.txt [hunk 1]
```diff
@@ -31,3 +31,13 @@
...
  */
+
+/**
+ * @defgroup    examples Examples
+ * @brief       Example applications
+ */
+
+/**
+ * @defgroup    tests Tests
+ * @brief       Tests applications
```
Comment (miri64) [inline doc.txt]: AFAIK those groups will be empty in the rendered doc. Do we really need them?

File: pkg/lwip/include/arch/cc.h [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 /**
  * @defgroup    lwip_arch_cc    Compiler and processor description
```
Comment (miri64) [inline pkg/lwip/include/arch/cc.h]: `pkg_lwip_arch_cc`?

File: pkg/lwip/include/arch/sys_arch.h [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 /**
  * @defgroup    lwip_arch_sys_arch Architecture depentent definitions
```
Comment (miri64) [inline pkg/lwip/include/arch/sys_arch.h]: `pkg_lwip_arch_sys_arch`?

File: pkg/lwip/include/lwipopts.h [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 /**
  * @defgroup    lwip_opts   lwIP options
```
Comment (miri64) [inline pkg/lwip/include/lwipopts.h]: `pkg_lwip_opts`?

File: sys/auto_init/auto_init.c [hunk 5]
```diff
@@ -7,7 +7,7 @@
  * General Public License v2.1. See the file LICENSE in the top level
  * directory for more details.
  *
- * @ingroup auto_init
+ * @ingroup sys_autoinit
```
Comment (miri64) [inline sys/auto_init/auto_init.c]: Here and below: shouldn't it rather be `sys_auto_init`?

File: sys/include/auto_init.h [hunk 6]
```diff
@@ -8,22 +8,55 @@
...
+
+/**
+ * @defgroup    sys_autoinit_gnrc_netif GNRC netif drivers auto-initialization
+ * @ingroup     sys_autoinit
+ * @brief       Provides auto-initialization of Netif drivers via GNRC.
+ *
+ * This module contains all netif drivers that support auto-initialization.
+ */
+
+/**
```
Comment (miri64) [inline sys/include/auto_init.h]: Unrelated change?

File: sys/include/crypto/modes/ccm.h [hunk 7]
```diff
@@ -26,11 +27,16 @@
...
+/**
+ * @name CCM error codes
+ * @{
+ */
+#define CCM_ERR_INVALID_NONCE_LENGTH        (-2)
+#define CCM_ERR_INVALID_CBC_MAC             (-3)
+#define CCM_ERR_INVALID_DATA_LENGTH         (-3)
+#define CCM_ERR_INVALID_LENGTH_ENCODING     (-4)
+#define CCM_ERR_INVALID_MAC_LENGTH          (-5)
+/** @} */
```
Comment (miri64) [inline sys/include/crypto/modes/ccm.h]: Unrelated

File: doc.txt (hunk 1 above)
Comment (aabadie) [reply doc.txt]: Yes, all examples are using `@ingroup examples` and tests `@ingroup tests`. I know this is not nice but at least it silents the check script.
There are 3 solutions:
1. Remove all `@ingroup tests` and `@ingroup examples`
2. Ignore undefined examples and tests groups in the check script
3. Define `tests` and `examples` groups and document them. If this solution is pushed a bit, we could have all examples and tests documented in doxygen (instead of READMEs).

I think `1.` should be avoided because it will touch too many files and maybe one day it can be useful to have those groups defined.
`2.` is the simplest.
`3.` is IMHO technically and from a centralized documentation the best solution. But requires more work.

File: pkg/lwip/include/arch/cc.h (hunk 2 above)
Comment (aabadie) [reply pkg/lwip/include/arch/cc.h]: Ok, but a bit out of the scope of this PR. If you insist, I can also change that.

File: sys/auto_init/auto_init.c (hunk 5 above)
Comment (aabadie) [reply sys/auto_init/auto_init.c]: I kept the initially defined `sys_autoinit`, I'll use your suggestion.

File: sys/include/auto_init.h (hunk 6 above)
Comment (aabadie) [reply sys/include/auto_init.h]: the changes related to auto_init was a bit problematic. I went for the improvement of the initial doc and add the missing groups there. Do you have a better suggestion ?

File: sys/include/crypto/modes/ccm.h (hunk 7 above)
Comment (aabadie) [reply sys/include/crypto/modes/ccm.h]: No, fixing the doxygen for this file is allowing the processing of doxygen in this file and, in the end, triggers a warning from doxygen because of undocumented defines. This change is fixing this. Do you want it in a separate commit ?

File: pkg/lwip/include/arch/cc.h (hunk 2 above)
Comment (miri64) [reply pkg/lwip/include/arch/cc.h]: It's less out of scope than the introduction of `auto_init` doc below since this PR is about group naming ;-).

Comment (aabadie) [reply pkg/lwip/include/arch/cc.h]: Ok, you insist ;)

File: doc.txt (hunk 1 above)
Comment (miri64) [reply doc.txt]: I think 3 does make little sense. While I prefer API documentation and usage documentation to be centralized, the READMEs of applications should be kept *with the application* for two reasons:

1. One can easily open a README in command-line or their favorite editor/IDE while working with the example and does not need to open an extra browser window
2. The context might get lost if the README is moved somewhere else

As such, I prefer 2, but instead of ignoring the *groups* I rather would ignore the *directories* (as e.g. the vendor headers).

Furthermore, I don't see the point of being so strict with group names in .c-files, as they are not parsed by doxygen, so they can't break the rendered doc (which was the whole point of my initial issue), so it might also be sensible to ignore .c-files altogether. But maybe I'm in the minority with this opinion.

File: sys/auto_init/auto_init.c (hunk 5 above)
Comment (aabadie) [reply sys/auto_init/auto_init.c]: name changed

File: pkg/lwip/include/lwipopts.h (hunk 4 above)
Comment (aabadie) [reply pkg/lwip/include/lwipopts.h]: all lwip pkg group are fixed

File: doc.txt (hunk 1 above)
Comment (aabadie) [reply doc.txt]: I dropped the defined examples and tests groups. Will update #9266 with solution 2.

File: sys/include/auto_init.h (hunk 6 above)
Comment (miri64) [reply sys/include/auto_init.h]: Ah sorry, forgot about this comment. Iff you insist to keep the doc improvement in this PR, please at least factor it out of https://github.com/RIOT-OS/RIOT/pull/9264/commits/6f91bf801bd3ede63fde923ea6fe4b027f4726cb into a separate commit.

Comment (aabadie) [reply sys/include/auto_init.h]: > please at least factor it out of 6f91bf8 into a separate commit.

done

**Comment by aabadie** [comment]
Rebased, now that #9266 is merged.

**Comment by aabadie** [comment]
@miri64, are your concerns addressed here so I can squash ?

File: sys/include/auto_init.h [hunk 8]
```diff
@@ -8,22 +8,55 @@
...
- *              initialized by auto_init. The modules will be initialized in
- *              the context of the main thread right before the main function
- *              gets called. Be aware that most modules expect to be
- *              initialized only once, so do not call a module's init function
- *              when using auto_init unless you know what you're doing.
  *
+ * This feature can be enabled in any application by adding the `auto_init`
+ * module to the application Makefile:
+ *
+ * ```mk
```
Comment (miri64) [inline sys/include/auto_init.h]: This syntax doesn't work in Doxygen. Try

```markdown
~~~~~~~~~~~~~~~~~~~~~~~ {.mk}
USEMODULE += auto_init
~~~~~~~~~~~~~~~~~~~~~~~
```

instead if you like to enable syntax highlighting, otherwise remove the `mk`.

File: sys/include/auto_init.h [hunk 9]
```diff
@@ -8,22 +8,55 @@
...
+ * USEMODULE += auto_init
+ * ```
+ *
+ * `auto_init` initializes any included module that provides auto
+ * initialization capabilities.
+ *
+ * Drivers or cpu peripherals that provides a SAUL adaption and net interfaces
```
Comment (miri64) [inline sys/include/auto_init.h]: "network interfaces"

File: sys/include/auto_init.h [hunk 10]
```diff
@@ -8,22 +8,55 @@
...
+ * can be initialized automatically with auto_init.
```
Comment (miri64) [inline sys/include/auto_init.h]: `` around `auto_init`

File: sys/include/auto_init.h [hunk 11]
```diff
@@ -8,22 +8,55 @@
...
+ *
+ * For high-level device drivers (@ref drivers), the default initialization
+ * parameters are taken from the `DRIVERS_params.h` files provided by the
```
Comment (miri64) [inline sys/include/auto_init.h]: `DRIVER`

File: sys/include/auto_init.h [hunk 12]
```diff
@@ -8,22 +8,55 @@
...
+ * driver implementation.
+ *
+ * From low-level CPU peripheral, the default initialization parameters are
+ * defined in each board configuration that supports them.
+ *
+ * The modules will be initialized in the context of the main thread right
+ * before the main function gets called. Be aware that most modules expect to
+ * be initialized only once, so do not call a module's init function when using
+ * auto_init unless you know what you're doing.
```
Comment (miri64) [inline sys/include/auto_init.h]:
```markdown
`auto_init`, unless you know what you're doing.
```

File: sys/include/auto_init.h [hunk 13]
```diff
@@ -8,22 +8,55 @@
...
+ */
+
+/**
+ * @defgroup    sys_auto_init_saul SAUL drivers auto-initialization
+ * @ingroup     sys_auto_init
+ * @brief       Provides auto-initialization of SAUL drivers.
+ *
+ * This module contains all high-level drivers (sensor or actuators) that
```
Comment (miri64) [inline sys/include/auto_init.h]: Please rephrase. This sentence makes little sense:

* What are high-level drivers? Do you mean SAUL as a high-level *interface* to the drivers?
* This module only allows for auto-initialization of those drivers using SAUL, and does not contain the drivers.

File: sys/include/auto_init.h [hunk 14]
```diff
@@ -8,22 +8,55 @@
...
+ * support auto-initialization via SAUL.
+ */
+
+/**
+ * @defgroup    sys_auto_init_gnrc_netif GNRC netif drivers auto-initialization
+ * @ingroup     sys_auto_init
+ * @brief       Provides auto-initialization of Netif drivers via GNRC.
+ *
+ * This module contains all netif drivers that support auto-initialization.
```
Comment (miri64) [inline sys/include/auto_init.h]: This module also doesn't contain any drivers, only enables their auto-initialization.

File: sys/include/auto_init.h [hunk 15]
```diff
@@ -8,22 +8,55 @@
...
+ * @brief       Provides auto-initialization of SAUL drivers.
+ *
+ * This module contains all high-level drivers (sensor or actuators) that
+ * support auto-initialization via SAUL.
+ */
+
+/**
+ * @defgroup    sys_auto_init_gnrc_netif GNRC netif drivers auto-initialization
+ * @ingroup     sys_auto_init
+ * @brief       Provides auto-initialization of Netif drivers via GNRC.
```
Comment (miri64) [inline sys/include/auto_init.h]: `s/Netif/network device drivers/`. Ideally, link to `drivers_netdev`, e.g.

```markdown
Provides auto-initialization for [network device drivers](@ref drivers_netdev) via GNRC.
```

File: sys/crypto/modes/cbc.c [hunk 16]
```diff
@@ -7,7 +7,7 @@
  */
 
 /**
- * @ingroup     sys_crypto_modes
+ * @ingroup     sys_crypto_mode_cbc
```
Comment (miri64) [inline sys/crypto/modes/cbc.c]: If you just use `sys_crypto` here, the introduction of the new groups below would not be required.

File: sys/include/crypto/modes/cbc.h [hunk 17]
```diff
@@ -7,11 +7,12 @@
  */
 
 /**
+ * @defgroup    sys_crypto_mode_cbc Cipher block chaining mode
```
Comment (miri64) [inline sys/include/crypto/modes/cbc.h]: This is really completely unrelated. These files are perfectly fine to all be grouped in `sys_crypto`. If you disagree, please move this group introductions to a separate PR.

File: sys/include/auto_init.h (hunk 8 above)
Comment (aabadie) [reply sys/include/auto_init.h]: removed the commit related to that change. Will open a dedicated PR for that later.

File: sys/crypto/modes/cbc.c (hunk 16 above)
Comment (aabadie) [reply sys/crypto/modes/cbc.c]: You are right, fixed now.

**Comment by aabadie** [comment]
@miri64, before this PR becomes unmaintainable, do you have other concerns ? I think I addressed all your comments and moved out the unrelated things.

**Comment by aabadie** [comment]
squashed, let's wait for the CI.

**Comment by miri64** [comment]
Travis fail is unrelated (and due to an old cppcheck version, I believe) => go
